openedx/core/lib/content_security_policy/middleware.py

```diff
@@ -0,0 +1,149 @@
+    #   Level 3 reports. If there are any semicolons or commas in the URL, they must be URL-encoded.
+    #   Level 3 reporting is only enabled if ``CSP_STATIC_REPORTING_NAME`` is also set.
+    reporting_uri = getattr(settings, 'CSP_STATIC_REPORTING_URI', None)
+
+    # .. setting_name: CSP_STATIC_REPORTING_NAME
+    # .. setting_default: None
+    # .. setting_description: Used for CSP Level 3 reporting. This sets the name to use in the
+    #   report-to CSP field and the Reporting-Endpoints header. If omitted, then Level 3 CSP
+    #   reporting will not be enabled. If present, this must be a string starting with an ASCII
+    #   letter and can contain ASCII letters, numbers, hyphen, underscore, and some other characters.
+    #   See https://www.rfc-editor.org/rfc/rfc8941.html#section-3.3.4 for full grammar.
+    reporting_endpoint_name = getattr(settings, 'CSP_STATIC_REPORTING_NAME', None)
+
+    if not enforce_policies and not report_policies:
+        return {}
+
+    headers = {}
+
+    reporting_suffix = ''
+    if reporting_uri:
+        reporting_suffix = f"; report-uri {reporting_uri}"
+        if reporting_endpoint_name:
+            headers['Reporting-Endpoints'] = f'{reporting_endpoint_name}="{reporting_uri}"'
+            reporting_suffix += f"; report-to {reporting_endpoint_name}"
+
+    def clean_header(value):
+        # Collapse any internal whitespace that contains a newline. This allows
+        # writing the setting value as a multi-line string, which is useful for
+        # CSP -- the values can be quite long.
+        value = re.sub("\\s*\n\\s*", " ", value).strip()
+        # Remove any trailing semicolon, which we allow (for convenience).
+        # The CSP spec does not allow trailing semicolons or empty directives.
+        value = re.sub("[;\\s]+$", "", value)
+        return value
+
+    if enforce_policies:
+        headers['Content-Security-Policy'] = clean_header(enforce_policies) + reporting_suffix
+
+    if report_policies:
+        headers['Content-Security-Policy-Report-Only'] = clean_header(report_policies) + reporting_suffix
+
+    return headers
+
+
+def _append_headers(response_headers, more_headers):
+    """
+    Append to the response headers. If a header already exists, assume it is
+    permitted to be multi-valued (comma-separated), and update the existing value.
+
+    Arguments:
+        response_headers: response.headers (or any dict-like object), to be modified
+        more_headers: Dict of header names to values
+    """
+    for k, v in more_headers.items():
+        if existing := response_headers.get(k):
+            response_headers[k] = f"{existing}, {v}"
+        else:
+            response_headers[k] = v
+
+
+class MiddlewareNotUsed(Exception):
+    """This middleware is not used in this server configuration"""
+
+
+def _get_override(request, url_specific_csps):
+    """
+    Check the CUSTOM_CSPS environment variable for a match to the request url.
```

Comment:
`Check the GET_CUSTOM_CSPS django configuration setting for a match to the request url.`

Reply:
updated